### PR TITLE
A use:portal action + misc new features

### PR DESCRIPTION
**Pull request by bfanger**
New features:
- Less DOM elements
- Reactive (changing target moves elements)
- target property is typed, provides typehints for typescript users.
- Portal contents are hidden in SSR render
- Functionality also available as action:`use:portal={'#target'}`
  - No wrapper div
  - Allows target and contents to be an inline or inline-block elements.
  - Allows for display: contents for grid and flex targets

Improved readme and added examples to demo/rollup-demo

### CODE REVIEW

**Comment by bfanger** [comment]
#27 adds unittests. nice! @johannchopin 
As there was some discussion, i've merged these into a different branch: https://github.com/bfanger/svelte-portal/tree/v3-with-tests

File: Portal.svelte [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+<script>
+  /**
+   * DOM Element or CSS Selector
+   * @type { HTMLElement|string}
+   */
+  export let target;
+</script>
+
+<span style="display: contents" use:portal={target}><slot /></span>
```
Comment (johannchopin) [inline Portal.svelte:43]: Render Portal children is a `span` isn't a good idea since it produces an invalid HTML structure.

Comment (bfanger) [reply Portal.svelte]: Agreed, using a portal for inline content is not common. and with the use:portal action `<span use:portal=".target">` still possible.
I've also removed `display: contents` as caniuse mentions buggy implementations. (and still possible with use:portal if you need it)

**Comment by romkor** [comment]
Thanks for PR!

> Breaking change:

>target prop is now required, no longer defaults to document.body
>Change <Portal> to <Portal target="body"> or <Portal target={document.body}> to get the previous behavior.
>import Portal from "svelte-portal/Portal.svelte" instead of import Portal from "svelte-portal" the former lacks editor support
Do not think we should make breaking changes.

Regarding whole PR will try to look at it during weekend.

File: package.json [hunk 2]
```diff
@@ -1,11 +1,8 @@
 {
   "name": "svelte-portal",
-  "version": "2.0.0",
+  "version": "3.0.0",
```
Comment (romkor) [inline package.json:3]: Version will be bumped by release-it

Comment (bfanger) [reply package.json:3]: I can revert the version change and add backward compatibility.

I'm currently maintaining 2 versions of the PR, one with and one without unittests.
If the unittest PR is merged first I can create a single PR on top of that.

I also like to add one additional feature: Hiding the content when in rendered in ssr mode. 

Looking forward to your full review.

Comment (romkor) [reply package.json:3]: @bfanger Please maintain only your PR.

File: src/Portal.svelte [hunk 3]
```diff
@@ -1,37 +1,62 @@
...
+   * @param {HTMLElement} el
+   * @param {HTMLElement|string} target DOM Element or CSS Selector
+   */
+  export function portal(el, target) {
+    let targetEl;
+    async function update(newTarget) {
+      target = newTarget;
+      if (typeof target === "string") {
+        targetEl = document.querySelector(target);
+        if (targetEl === null) {
```
Comment (romkor) [inline src/Portal.svelte:16]: Do we really need this `if` block?

File: src/Portal.svelte [hunk 4]
```diff
@@ -1,37 +1,62 @@
...
-      // Force exit
-      if (targetEl === null) {
-        return () => {};
+  /**
+   * Usage: <div use:portal={'css selector'}> or <div use:portal={document.body}>
+   *
+   * @param {HTMLElement} el
+   * @param {HTMLElement|string} target DOM Element or CSS Selector
+   */
+  export function portal(el, target) {
```
Comment (romkor) [inline src/Portal.svelte:10]: We can add default target to let directive work 'out of the box' 
`export function portal(el, target = "body")`

File: src/Portal.svelte (hunk 3 above)
Comment (bfanger) [reply src/Portal.svelte:16]: In the unittests the target element was also rendered by svelte, i think this is a valid usecase, but then the element is not yet in the dom, waiting one tick and trying again will then find the freshly rendered element

File: src/Portal.svelte (hunk 4 above)
Comment (bfanger) [reply src/Portal.svelte]: Fixed

**Comment by romkor** [comment]
Hi @bfanger, can you merge latest master and push it.

**Comment by bfanger** [comment]
11a1cdc are the changed rebased on latest master, i've also added a `.prettierignore` to exclude the demo build directory.

Running tests locally shows `Tests:       4 passed, 4 total` not sure why the PR says: _Waiting for status to be reported_

**Comment by romkor** [comment]
> i've also added a .prettierignore to exclude the demo build directory.

But why? Point is to make all code base (source, demos, tests) look the same.

**Comment by johannchopin** [comment]
@romkor Why do we have demos for webpack and rollup? This project is a simple module and should be used the same way in both environment. I guess a demo in the README is enough 🤔

**Comment by romkor** [comment]
> @romkor Why do we have demos for webpack and rollup? This project is a simple module and should be used the same way in both environment. I guess a demo in the README is enough 🤔

I guess it is not enough. `demos` is good minimal starting point/sandbox.

**Comment by johannchopin** [comment]
@romkor Sure it's not enough but it would be probably better to show this demos in online svelte REPL. Because as a user I don't want to clone the repo, install the packages and start a build just to see some demos 👍

**Comment by romkor** [comment]
> @romkor Sure it's not enough but it would be probably better to show this demos in online svelte REPL. Because as a user I don't want to clone the repo, install the packages and start a build just to see some demos 👍

It's personal preferences. There are also REPL in about section https://svelte.dev/repl/407576d4fa984cfb97dcdd3da98e833e

**Comment by romkor** [comment]
> > i've also added a .prettierignore to exclude the demo build directory.
> 
> But why? Point is to make all code base (source, demos, tests) look the same.

Now I see, you have added only build directory to .prettierignore.

**Comment by romkor** [comment]
@bfanger Thank you for contribution 👍

**Pull request merged**